app/__init__.py

```diff
@@ -0,0 +1,109 @@
+"""
+A small webapp that functions on MountainProject.com data.
+
+It imports data based on a user ID, then analyzes and displays it.
+Code by Zach Wahrer.
+"""
+
+import re
+
+from flask import Flask, render_template, request, redirect
+from .config import *
+
+from .helpers import get_userid, ticklist, db_load, db_connect, db_close
+from .graphing import height_climbed, pitches_climbed, grade_scatter, get_types
+
+
+def create_app(test_config=None):
```

Comment:
This restructure kind of borked my production configuration. I can get MPV to run (both in my dev environment and on the production server) if I manually start Flask (as you noted with `flask run`). Unfortunately, when I try to run it using gunicorn (after changing my wsgi config file), it thinks I'm trying to pass values into `create_app()` rather than `app.run()`.

I spent a bunch of time today messing with the wsgi config, gunicorn, and looking for other flask implementations that use a similar layout, but unfortunately, everything I could find has `app = Flask(__name__)` outside of a function.

I figured out a workaround where I just removed  the `create_app()` def and the `test_config` if/else on the server code. I'd rather not have to do that every time I update the code there, but for now, I can't figure out a better way. If you have any thoughts, I'd love to hear them.

I'll probably try to dig into this another day, when I have time. I'm not saying your code is bad or that I want to revert, just that it is a bit hard to implement with gunicorn. :-) Anyway, no biggie. Just learning how to communicate and work remotely with someone. Thanks again for helping out with this project.

Reply:
Yeah I had suspected this might effect your production environment, sorry that ended up being the case but I'm glad you were able to find a work around! I did a bit of searching for running an app factory and configuring for deployment. I'm not sure this is much help but I found this resource:

This one is gunicorn specific
https://stackoverflow.com/questions/25319690/how-do-i-run-a-flask-app-in-gunicorn-if-i-used-the-application-factory-pattern

I think this was apache specific but the logic looked to be the same.
https://stackoverflow.com/questions/36499982/how-to-deploy-a-flask-app-with-factory-function-for-application-creation

The common thread seems to be that you need to point your configuration file to an application object, and that application object can be a call to `create_app()`. 

Also from stackoverflow:

According to the Flask Documentation, an application factory is good because:

- Testing. You can have instances of the application with different settings to test every case.

- Multiple instances. Imagine you want to run different versions of the same application. Of course you could have multiple instances with different configs set up in your webserver, but if you use factories, you can have multiple instances of the same application running in the same application process which can be handy

Ultimately this was the reason for adding an application factory. So we could test against the actual MPV app.

Let me know if I can be of any help answering any other questions!

Reply:
Perfect! The gunicorn specific link had just what I needed. I didn't know this was an app factory setup, so that was what was throwing me off knowing what to search for. Such an easy fix once I had the right info. Thanks for passing it along!